### PR TITLE
8286197: C2: Optimize MemorySegment shape in int loop

### PR DESCRIPTION
**Pull request by rwestrel**
This is another small enhancement for a code shape that showed up in a
MemorySegment micro benchmark. The shape to optimize is the one from test1:

```
for (int i = 0; i < size; i++) {
    long j = i * UNSAFE.ARRAY_INT_INDEX_SCALE;

    j = Objects.checkIndex(j, size * 4);

    if (((base + j) & 3) != 0) {
        throw new RuntimeException();
    }

    v += UNSAFE.getInt(base + j);
}
```

In that code shape, the loop iv is first scaled, result is then casted
to long, range checked and finally address of memory location is
computed.

The alignment check is transformed so the loop body has no check In
order to eliminate the range check, that loop is transformed into:

```
for (int i1 = ..) {
    for (int i2 = ..) {
        long j = (i1 + i2) * UNSAFE.ARRAY_INT_INDEX_SCALE;

        j = Objects.checkIndex(j, size * 4);

        v += UNSAFE.getInt(base + j);
    }
}
```

The address shape is (AddP base (CastLL (ConvI2L (LShiftI (AddI ...

In this case, the type of the ConvI2L is [min_jint, max_jint] and type
of CastLL is [0, max_jint] (the CastLL has a narrower type).

I propose transforming (CastLL (ConvI2L into (ConvI2L (CastII in that
case. The convI2L and CastII types can be set to [0, max_jint]. The
new address shape is then:

(AddP base (ConvI2L (CastII (LShiftI (AddI ...

which optimize well.

(LShiftI (AddI ...
is transformed into
(AddI (LShiftI ...
because one of the AddI input is loop invariant (i2) and we have:

(AddP base (ConvI2L (CastII (AddI (LShiftI ...

Then because the ConvI2L and CastII types are [0, max_jint], the AddI
is pushed through the ConvI2L and CastII:

(AddP base (AddL (ConvI2L (CastII (LShiftI ...

base and one of the inputs of the AddL are loop invariant so this
transformed into:

(AddP (AddP ...) (ConvI2L (CastII (LShiftI ...

The (AddP ...) is loop invariant so computed before entry. The
(ConvI2L ...) only depends on the loop iv.

The resulting address is a shift + an add. The address before
transformation requires 2 adds + a shift. Also after unrolling, the
adress of the second access in the loop is cheaper to compute as it
can be derived from the address of the first access.

For all of this to work:
1) I added a CastLL::Ideal transformation:
(CastLL (ConvI2L into (ConvI2l (CastII

2) I also had to prevent split if to transform (LShiftI (Phi for the
iv Phi of a counted loop.

test2 and test3 test 1) and 2) separately.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8286197](https://bugs.openjdk.org/browse/JDK-8286197): C2: Optimize MemorySegment shape in int loop

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [a122f0cf](https://git.openjdk.org/jdk/pull/8555/files/a122f0cf7aaed64b0fbcc6b1984de2f81ccb8516)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/8555/head:pull/8555` \
`$ git checkout pull/8555`

Update a local copy of the PR: \
`$ git checkout pull/8555` \
`$ git pull https://git.openjdk.org/jdk pull/8555/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8555`

View PR using the GUI difftool: \
`$ git pr show -t 8555`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/8555.diff">https://git.openjdk.org/jdk/pull/8555.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/castnode.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/castnode.cpp:385]: `t != t_in` does not mean that type is narrower in general case. I think we need to check ranges (types meet?).

**Review by vnkozlov — COMMENTED** [review]
Good suggestion. I have comments.

File: src/hotspot/share/opto/castnode.cpp
Comment (vnkozlov) [reply src/hotspot/share/opto/castnode.cpp:385]: There is no specialized `CastLLNode::Value()` and `ConstraintCastNode` only calls `filter_speculative()` which do call `join()`. May be it is indeed enough. Yes, would be nice to have an assert to make sure we got it right.

Comment (rwestrel) [reply src/hotspot/share/opto/castnode.cpp]: The new commit adds an assert

File: src/hotspot/share/opto/loopopts.cpp
Comment (rwestrel) [reply src/hotspot/share/opto/loopopts.cpp]: fixed in new commit

**Review by vnkozlov — APPROVED** [review]
Good. I will start testing.

**Comment by vnkozlov** [comment]
Testing passed. It needs second review.

**Comment by rwestrel** [comment]
Any one else for this one?

**Review by TobiHartmann — APPROVED** [review]
Looks good to me but I'm wondering if we should delay this to JDK 20 as we are late for the JDK 19 release and had many issues with Cast/ConvNodes before.

**Review by TobiHartmann — APPROVED** [review]

**Pull request closed without merging**
